pkg/steps/pre_publish_images.go

```diff
@@ -0,0 +1,121 @@
+
+	fromImage := "dry-fake"
+	if !dry {
+		from, err := s.istClient.ImageStreamTags(s.jobSpec.Namespace).Get(fmt.Sprintf("%s:%s", api.PipelineImageStream, s.config.From), meta.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("could not resolve base image: %v", err)
+		}
+		fromImage = from.Image.Name
+	}
+
+	is := newImageStream(s.config.To.Namespace, s.config.To.Name)
+	ist := s.imageStreamTag(fromImage)
+
+	return createImageStreamWithTag(s.isClient, s.istClient, is, ist, dry)
+}
+
+func (s *prePublishOutputImageTagStep) Done() (bool, error) {
+	log.Printf("Checking for existence of %s/%s:%s", s.config.To.Namespace, s.config.To.Name, s.tag)
+
+	ist, err := s.istClient.ImageStreamTags(s.config.To.Namespace).Get(fmt.Sprintf("%s:%s", s.config.To.Name, s.tag), meta.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("could not retrieve output imagestreamtag: %v", err)
+	}
+
+	// TODO(chance): this doesn't handle dry run since Done() doesn't have
+	// information about if it's a dry-run
+	from, err := s.istClient.ImageStreamTags(s.jobSpec.Namespace).Get(fmt.Sprintf("%s:%s", api.PipelineImageStream, s.config.From), meta.GetOptions{})
+	if err != nil {
+		return false, fmt.Errorf("could not resolve base image: %v", err)
+	}
+
+	desiredIst := s.imageStreamTag(from.Image.Name)
+	// if a tag already exists but doesn't match what we're looking for we're
+	// not done
+	return equality.Semantic.DeepEqual(ist.Tag, desiredIst.Tag), nil
+}
+
+func (s *prePublishOutputImageTagStep) Requires() []api.StepLink {
+	return []api.StepLink{api.InternalImageLink(s.config.From), api.ReleaseImagesLink()}
+}
+
+func (s *prePublishOutputImageTagStep) Creates() []api.StepLink {
+	ref := api.ImageStreamTagReference{
+		Name:      s.config.To.Name,
+		Tag:       s.tag,
+		Namespace: s.config.To.Namespace,
+	}
+	return []api.StepLink{api.ExternalImageLink(ref)}
+}
+
+func (s *prePublishOutputImageTagStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+
+func (s *prePublishOutputImageTagStep) Name() string {
+	return fmt.Sprintf("[prepublish:%s:%s]", s.config.To.Namespace, s.config.To.Name)
+}
+
+func (s *prePublishOutputImageTagStep) Description() string {
+	return fmt.Sprintf("Tag the image %s into the image stream tag %s/%s:%s", s.config.From, s.config.To.Namespace, s.config.To.Name, s.tag)
+}
+
+func (s *prePublishOutputImageTagStep) imageStreamTag(fromImage string) *imageapi.ImageStreamTag {
+	return newImageStreamTag(
+		s.jobSpec.Namespace, api.PipelineImageStream, fromImage,
+		s.config.To.Namespace, s.config.To.Name, s.tag,
+	)
+}
+
+func PrePublishOutputImageTagStep(config api.PrePublishOutputImageTagStepConfiguration, istClient imageclientset.ImageStreamTagsGetter, isClient imageclientset.ImageStreamsGetter, jobSpec *api.JobSpec) api.Step {
+	pull := jobSpec.Refs.Pulls[0]
+	pullNumber := strconv.Itoa(pull.Number)
+	tag := "pr-" + pullNumber
```

Comment:
This will be confusing if a job with prepublish is ever run by our rehearse tool. Those jobs will have a `pull` that correspond to a PR to the `openshift/release` repo, but that information is lost here, under assumption that the org/repo here will always be the "home" repo. In the worst (but rare) case you could feed images from two PRs (one to the home repo and one to openshift/release) into the same tag. To be on the safe side we could incorporate the org-repo parts to the tag name.

Rehearsing jobs with side effects is tricky anyway - the tool was built under an assumption that presubmits do not have side effects. We'll need to think about that interaction a bit.

Reply:
True. How does this get handled by other image oriented steps? Do we need to expose the fact that something is a rehearse, and allow steps to inspect that from the jobSpec so the step can skip doing certain things under rehearse?

Reply:
Slack convo said we should ignore this for now, right?

Reply:
IMO yes